perf(linter): store rules by AST type in a boxed array

overlookmotel opened the pull request:
Follow-on after #13138.

`rules_by_ast_type` has a fixed length, known at compile time. So use a boxed array for it, instead of a `Vec`.

This has 2 advantages:

1. `Box<[Vec<T>; N]>` is 8 bytes, instead of 24 for `Vec<Vec<T>>`.

2. The length of the array is encoded in the type, so it may allow compiler to deduce that indexing into it with an `AstType` cannot go out of bounds, and remove bounds checks (maybe - if compiler is smart enough, it may already be able to deduce that with the `Vec`).

overlookmotel commented on crates/oxc_linter/src/lib.rs:
```diff
@@ -169,13 +169,23 @@ impl Linter {
+                // cannot go out of bounds, and remove bounds checks. The `unwrap` is infallible and should be optimized out.
+                let rules_by_ast_type = vec![Vec::new(); AST_TYPES_LEN];
+                #[expect(clippy::missing_panics_doc, reason = "infallible")]
+                let mut rules_by_ast_type =
+                    Box::<[_; AST_TYPES_LEN]>::try_from(rules_by_ast_type.into_boxed_slice())
+                        .ok()
+                        .unwrap();
```
Without `.ok()`, does not compile, because `LintContext` is not `Debug`.

camc314 commented:
### Merge activity

* **Sep 8, 3:20 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(linter): store rules by AST type in a boxed array (#13578)

Follow-on after #13138.

`rules_by_ast_type` has a fixed length, known at compile time. So use a boxed array for it, instead of a `Vec`.

This has 2 advantages:

1. `Box<[Vec<T>; N]>` is 8 bytes, instead of 24 for `Vec<Vec<T>>`.

2. The length of the array is encoded in the type, so it may allow compiler to deduce that indexing into it with an `AstType` cannot go out of bounds, and remove bounds checks (maybe - if compiler is smart enough, it may already be able to deduce that with the `Vec`).
